### DISCUSSION TITLE
Error using curl model

### DISCUSSION THREAD

**Discussion by MYMahfouz** [OpenFAST/openfast#1721] [General]
Dear Jason,

I am receiving the following error when trying to use the curl model.

I am using the same models that already worked with the polar model just setting `Mod_Wake = 2` instead of 1 for the polar implementation.

Am I overlooking something? Should I update other inputs for the curl model?

````
 Time: 19 of 4200 seconds.  Estimated final completion at 01:34:25.

FARM_CalcOutput:AWAE_CalcOutput:LowResGridCalcOutput:The center of wake plane #1 for turbine #1
has passed the lowest-most X boundary of the low-resolution domain.
LowResGridCalcOutput:The center of wake plane #1 for turbine #1 has passed the lowest-most Y
boundary of the low-resolution domain.
LowResGridCalcOutput:The center of wake plane #1 for turbine #1 has passed the upper-most Z
boundary of the low-resolution domain.
LowResGridCalcOutput:The center of wake plane #2 for turbine #1 has passed the lowest-most X
boundary of the low-resolution domain.
LowResGridCalcOutput:The center of wake plane #2 for turbine #1 has passed the lowest-most Y
boundary of the low-resolution domain.
LowResGridCalcOutput:The center of wake plane #2 for turbine #1 has passed the upper-most Z
boundary of the low-resolution domain.
LowResGridCalcOutput:The center of wake plane #3 for turbine #1 has passed the lowest-most X
boundary of the low-resolution domain.
LowResGridCalcOutput:The center of wake plane #3 for turbine #1 has pa

                                                                                                  
 Time: 23 of 4200 seconds.  Estimated final completion at 01:16:51.
Warning: SkewedWakeCorrection encountered a large value of chi (-174.7 deg), so the yaw
correction will be limited. This warning will not be repeated though the condition may persist.
See the AD15 chi output channels, and consider turning off the Pitt/Peters skew model (set
SkewMod=1) if this condition persists.

The BEM solution is being turned off due to low TSR.  (TSR = -2.30740E-05). This warning will not
be repeated though the condition may persist. (See GeomPhi output channel.)

T9:FARM_UpdateStates:FWrap_Increment:FAST_Solution:CalcOutputs_And_SolveForInputs:SolveOption2:AD_
CalcOutput:RotCalcOutput:BEMT_CalcOutput(node 37, blade 1):UA_CalcOutput:Mach number exceeds 0.3.
Theory is invalid. This warning will not be repeated though the condition may persist.
BEMT_CalcOutput(node 38, blade 1):UA_CalcOutput:Mach number exceeds 1.0. Equations cannot be
evaluated.


````

Best regards

**Comment by jjonkman** [comment]
Dear @MYMahfouz,

If all you've changed in your FAST.Farm input is `Mod_Wake`, from 1 to 2, my guess is that you are not following the modeling guidance for curled wake.  In particular, when using `Mod_Wake` = 2, `DT_Low` is tied to `dr`, see: https://openfast.readthedocs.io/en/main/source/user/fast.farm/ModelGuidance.html#low-resolution-domain.  (Using a larger value of `DT_Low` will result in numerically instability of the curled wake model.)  Please note, though, that `dr` can likely be bigger than you are using, e.g., `dr` = D/10 or D/15, as documented here: https://openfast.readthedocs.io/en/main/source/user/fast.farm/ModelGuidance.html#wake-dynamics-parameters.  (This same guidance on `dr` should actually apply to the polar model as well, but we still have to correct this section of the FAST.Farm modeling guidance.)

Best regards,

**Reply by MYMahfouz** [reply]
Dear @jjonkman,

I have done these changes and reran my model. I had 32 cases running over night and now i see 15 cases failed.

The errors I am getting are either the error I had before or these errors below. This is not the 3.5.1 version but rather v3.4.1 that I am using.

What else should i check beside changing `DT_Low` and `dr`? Do I need to decrease the time step further? my current `DT_Low=0.84`.

Thanks for all the help and support.

````
T5:FARM_UpdateStates:NearWakeCorrection:Wake model is not valid in the propeller-brake region,
i.e., Ct > 2.0.
````


````
T4:FARM_UpdateStates:FWrap_Increment:FAST_Solution:FAST_AdvanceStates:SolveOption2c_Inp2AD_SrvD:In
flowWind_CalcOutput:CalcOutput:IfW_4Dext_CalcOutput [position=(NaN, NaN, NaN) in wind-file
coordinates]:Interp4D:Outside the grid bounds. lower bounds
InflowWind_CalcOutput:InflowWind_GetSpatialAverage:CalcOutput:IfW_4Dext_CalcOutput
[position=(NaN, NaN, NaN) in wind-file coordinates]:Interp4D:Outside the grid bounds. lower bounds
InflowWind_CalcOutput:InflowWind_GetSpatialAverage:CalcOutput:IfW_4Dext_CalcOutput
[position=(NaN, NaN, NaN) in wind-file coordinates]:Interp4D:Outside the grid bounds. lower bounds
````

Best regards,

**Reply by jjonkman** [reply]
Dear @MYMahfouz,

I'm guessing that your model is on the edge of instability such that some cases run fine and some don't.  I would suggest reducing `DT_Low` a bit more and see if that solves the issue.

Best regards,

**Reply by MYMahfouz** [reply]
Dear @jjonkman,

I have compiled a binary version of FAST.Farm to use on linux. However, when i ran the exact same model I used for running the windows executable the curl model was unstable. I tried changing the `DT_Low` but it didn't help. I had to increase the `NumRadii` to fix this issue. 
Is it normal to have differences between the executable and linux binary versions since the compilers are different? Does this change of `NumRadii` make sense?

 Best regards,

**Reply by jjonkman** [reply]
Dear @MYMahfouz,

I would normally expect OpenFAST and FAST.Farm to produce very similar results across different operating systems and compilers.

Is the executable you compiled on Linus compiled with the same precision as compiled on Windows?

Best regards,

**Reply by MYMahfouz** [reply]
Dear @jjonkman,

The outputs are the same if I increase `NumRadii`, if i don't the linux version with the curl model gets unstable. In the executable version i use a value of 16 and in the linux version i use a value of 30. Would you say this is normal? Or is it something i should worry about.

Both versions are single precision.


Best regards,

**Reply by jjonkman** [reply]
I'm not heard of this issue before.

@ebranlard -- Any ideas?

**Reply by MYMahfouz** [reply]
Dear @jjonkman 
Would you in general expect two identical input files to work in the executable and not in linux?

I am not sure if my linux binary is the problem, but once i get it to work the results are identical.

**Reply by MYMahfouz** [reply]
I also sometimes get this error and restarting the simulation fixes it. The error appear right after the Mann box is read and before all time steps.

> Farm_Initialize:AWAE_Init:AWAE_UpdateStates:InflowWind_CalcOutput:InflowWind_GetSpatialAverage:Cal
 cOutput:IfW_FFWind_CalcOutput [position=(NaN, NaN, NaN)  in wind-file coordinates]: FF wind array
 boundaries violated: Grid too small in Y direction. Y=NaN; Y boundaries = [-2557.5, 2557.5]

**Reply by jjonkman** [reply]
I would not expect that running multiple simulations with the same input files would produce different results.  Are you running FAST.Farm with OpenMP parallelization?

@ebranlard -- Any ideas?

**Reply by MYMahfouz** [reply]
I am not using OpenMP.

Just a quick explanation on why i started playing with `NumRadii` values. As the figure here this is at the beginning of simulation. There is this unstable wake behind the turbine. The width of this instability is the `NumRadii` or at least that's what i think.

![image](https://github.com/OpenFAST/openfast/assets/33868271/2594a1a2-9a26-4a44-ab99-096061ac955b)


After few seconds this disappears and only a part of it, that i circled remains.

![image](https://github.com/OpenFAST/openfast/assets/33868271/717cfe5f-7d8c-4854-9e18-c0b100fba68a)


From what i saw this changes a lot with changing  `NumRadii`.


Another observation that makes me think something is weird is that this always happens only behind the first turbine in the .fstf file. At the beginning i thought maybe the wind field is corrupted, but this happens for all wind directions in which i rotate the turbines and not the wind field. Therefore, the first turbine is located at different positions of the wind field and still this happens in all wind directions.

These observations i made with the binary for linux i didnt check if similar observations can be made with the windows executable.

**Reply by MYMahfouz** [reply]
Sorry for spamming, but one last thing that i forgot to mention cause it was not so important for me but maybe it can help you tell if there is a difference between the difference in compilation.

For the inflowind input file:
In the executable on windows for the values of of `nx`, `ny`, and `nz`, I had floats so for example my `nz` was `128.0`.

In the linux binary this caused an error and i had to use integers so i switched the value to `128`.

I thought it is not that important but maybe this means there are more differences in how the inflowind module is compiled?

**Reply by ebranlard** [reply]
I don't have much idea of what can be going on... I haven't seen these issues before.
